File: app/src/main/java/jp/juggler/subwaytooter/table/NotificationTracking.java
package jp.juggler.subwaytooter.table;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.provider.BaseColumns;

import jp.juggler.subwaytooter.App1;
import jp.juggler.subwaytooter.util.LogCategory;

public class NotificationTracking {
	
	private static final LogCategory log = new LogCategory( "NotificationTracking" );
	
	private static final String table = "noti_trac";
	
	private static final String COL_ID = BaseColumns._ID;
	
	// アカウントDBの行ID。 サーバ側のIDではない
	private static final String COL_ACCOUNT_DB_ID = "a";
	
	// サーバから通知を取得した時刻
	private static final String COL_LAST_LOAD = "ll";
	
	// サーバから最後に読んだデータ。既読は排除されてるかも
	private static final String COL_LAST_DATA = "ld";
	
	// 通知ID。ここまで既読
	private static final String COL_NID_READ = "nr";
	
	// 通知ID。もっとも最近取得したもの
	private static final String COL_NID_SHOW = "ns";
	
	// 最後に表示した通知のID
	private static final String COL_POST_ID = "pi";
	
	// 最後に表示した通知の作成時刻
	private static final String COL_POST_TIME = "pt";
	
	public static void onDBCreate( SQLiteDatabase db ){

		db.execSQL(
			"create table if not exists " + table
				+ "(_id INTEGER PRIMARY KEY"
				+ ",a integer not null"
				+ ",ll integer default 0"
				+ ",ld text"
				+ ",nr integer default 0"
				+ ",ns integer default 0"
				+ ",pi integer default 0"
				+ ",pt integer default 0"
				+ ")"
		);
		db.execSQL(
			"create unique index if not exists " + table + "_a on " + table + "(a)"
		);
	}
	
	public static void onDBUpgrade( SQLiteDatabase db, int oldVersion, int newVersion ){
		if( oldVersion < 2 && newVersion >= 2 ){
			onDBCreate( db );
		}
	}
	
	private long id = -1L;
	private long account_db_id;
	public long last_load;
	public long nid_read;
	public long nid_show;
	
	public long post_id;
	public long post_time;
	
	public String last_data;
	
	private static final String WHERE_AID = COL_ACCOUNT_DB_ID + "=?";
	
	public static NotificationTracking load( long account_db_id ){
		NotificationTracking dst = new NotificationTracking();
		dst.account_db_id = account_db_id;
		try{
			Cursor cursor = App1.getDB().query( table, null,WHERE_AID, new String[]{ Long.toString( account_db_id ) }, null, null, null );
			try{
				if( cursor.moveToFirst() ){
					dst.id = cursor.getLong( cursor.getColumnIndex( COL_ID ) );
					dst.last_load = cursor.getLong( cursor.getColumnIndex( COL_LAST_LOAD ) );
					dst.nid_read = cursor.getLong( cursor.getColumnIndex( COL_NID_READ ) );
					dst.nid_show = cursor.getLong( cursor.getColumnIndex( COL_NID_SHOW ) );
					
					dst.post_id = cursor.getLong( cursor.getColumnIndex( COL_POST_ID ) );
					dst.post_time = cursor.getLong( cursor.getColumnIndex( COL_POST_TIME ) );
					
					int idx_last_data = cursor.getColumnIndex( COL_LAST_DATA );
					dst.last_data = cursor.isNull( idx_last_data ) ? null : cursor.getString( idx_last_data );

					log.d("load account_db_id=%s,post=%s,%s last_data=%s"
						,account_db_id
						,dst.post_id
						,dst.post_time
						,(dst.last_data ==null ? "null" : ""+dst.last_data.length())
					);
				}
			}finally{
				cursor.close();
			}
		}catch( Throwable ex ){
			log.e( ex, "load failed." );
		}
		return dst;
	}
	
	public void save(){
		try{
			ContentValues cv = new ContentValues();
			cv.put( COL_ACCOUNT_DB_ID, account_db_id );
			cv.put( COL_LAST_LOAD, last_load );
			cv.put( COL_NID_READ, nid_read );
			cv.put( COL_NID_SHOW, nid_show );
			cv.put( COL_LAST_DATA, last_data );
			cv.put( COL_POST_ID, post_id);
			cv.put( COL_POST_TIME, post_time);
			if( id == -1L ){
				id = App1.getDB().insert( table,null,cv );
				log.d("save.insert account_db_id=%s,post=%s,%s last_data=%s"
					,account_db_id
					,post_id
					,post_time
					,(last_data ==null ? "null" : ""+last_data.length())
				);
				
			}else{
				App1.getDB().update( table, cv, WHERE_AID, new String[]{ Long.toString( account_db_id ) } );
				log.d("save.update account_db_id=%s,post=%s,%s last_data=%s"
					,account_db_id
					,post_id
					,post_time
					,(last_data ==null ? "null" : ""+last_data.length())
				);
			}
		}catch( Throwable ex ){
			log.e( ex, "save failed." );
		}
	}
	
	public void updatePost(long post_id,long post_time){
		this.post_id = post_id;
		this.post_time = post_time;
		try{
			ContentValues cv = new ContentValues();
			cv.put( COL_POST_ID, post_id );
			cv.put( COL_POST_TIME, post_time );
			int rows = App1.getDB().update( table, cv,WHERE_AID, new String[]{ Long.toString( account_db_id ) } );
			log.d("updatePost account_db_id=%s,post=%s,%s last_data=%s,update_rows=%s"
				,account_db_id
				,post_id
				,post_time
				,(last_data ==null ? "null" : ""+last_data.length())
				,rows
			);
		}catch( Throwable ex ){
			log.e( ex, "updatePost failed." );
		}
	}
	
	public static void updateRead(long account_db_id){
		try{
			String[] where_args = new String[]{ Long.toString( account_db_id ) };
			Cursor cursor = App1.getDB().query( table, new String[]{ COL_NID_SHOW,COL_NID_READ }, WHERE_AID, where_args, null, null, null );
			try{
				if( cursor.moveToFirst() ){
					long nid_show = cursor.getLong( cursor.getColumnIndex( COL_NID_SHOW ) );
					long nid_read = cursor.getLong( cursor.getColumnIndex( COL_NID_READ ) );
					log.d("updateRead account_db_id=%s, nid_show=%s, nid_read=%s",account_db_id,nid_show,nid_read);
					ContentValues cv = new ContentValues();
					cv.put( COL_NID_READ, nid_show );
					App1.getDB().update( table, cv, WHERE_AID,where_args );
				}else{
					log.e("updateRead no row in query.");
				}
			}finally{
				cursor.close();
			}
		}catch( Throwable ex ){
			log.e( ex, "updateRead failed." );
		}
	}
	
	public static void resetPostAll(){
		try{
			ContentValues cv = new ContentValues();
			cv.put( COL_POST_ID, 0 );
			cv.put( COL_POST_TIME, 0 );
			App1.getDB().update( table, cv,null,null);
			
		}catch( Throwable ex ){
			log.e( ex, "resetPostAll failed." );
		}
	}
	
	public static void resetLastLoad(long db_id){
		try{
			ContentValues cv = new ContentValues();
			cv.put( COL_LAST_LOAD, 0 );
			App1.getDB().update( table, cv, COL_ACCOUNT_DB_ID+"=?",new String[]{ Long.toString( db_id )});
		}catch( Throwable ex ){
			log.e( ex, "resetLastLoad(db_id) failed." );
		}
	}

	public static void resetLastLoad(){
		try{
			ContentValues cv = new ContentValues();
			cv.put( COL_LAST_LOAD, 0 );
			App1.getDB().update( table, cv,null,null);
		}catch( Throwable ex ){
			log.e( ex, "resetLastLoad() failed." );
		}
	}
}

File: app/src/main/java/jp/juggler/subwaytooter/table/NotificationTracking.kt
package jp.juggler.subwaytooter.table

import android.content.ContentValues
import android.database.sqlite.SQLiteDatabase
import android.provider.BaseColumns
import jp.juggler.subwaytooter.api.entity.Acct
import jp.juggler.subwaytooter.api.entity.EntityId
import jp.juggler.subwaytooter.api.entity.putMayNull
import jp.juggler.util.data.*
import jp.juggler.util.log.LogCategory
import java.util.concurrent.ConcurrentHashMap

class NotificationTracking {

    private var dirty = false

    private var id = -1L
        set(value) {
            dirty = true; field = value
        }

    // アカウントDBの行ID。 サーバ側のIDではない
    private var accountDbId: Long = 0
        set(value) {
            dirty = true; field = value
        }

    // 返信だけ通知グループを分ける
    private var notificationType: String = ""
        set(value) {
            dirty = true; field = value
        }

    // 通知ID。ここまで既読
    var nid_read: EntityId? = null
        set(value) {
            dirty = true; field = value
        }

    // 通知ID。もっとも最近取得したもの
    var nid_show: EntityId? = null
        set(value) {
            dirty = true; field = value
        }

    // 最後に表示した通知のID
    var post_id: EntityId? = null
        set(value) {
            dirty = true; field = value
        }

    // 最後に表示した通知の作成時刻
    var post_time: Long = 0
        set(value) {
            dirty = true; field = value
        }

    companion object : TableCompanion {
        private val log = LogCategory("NotificationTracking")
        override val table = "noti_trac"
        private const val COL_ID = BaseColumns._ID
        private const val COL_ACCOUNT_DB_ID = "a"
        private const val COL_NID_READ = "nr"
        private const val COL_NID_SHOW = "ns"
        private const val COL_POST_ID = "pi"
        private const val COL_POST_TIME = "pt"
        private const val COL_NOTIFICATION_TYPE = "nt"

        private val metaColumns = MetaColumns(table, initialVersion = 0).apply {
            column(0, COL_ID, "INTEGER PRIMARY KEY")
            column(0, COL_ACCOUNT_DB_ID, "integer not null")
            column(0, COL_NID_READ, "text")
            column(0, COL_NID_SHOW, "text")
            column(0, COL_POST_ID, "text")
            column(0, COL_POST_TIME, "integer default 0")
            column(40, COL_NOTIFICATION_TYPE, "text default ''")
            deleteBeforeCreate = true
            createExtra = {
                arrayOf(
                    "create unique index if not exists ${table}_b on $table($COL_ACCOUNT_DB_ID,$COL_NOTIFICATION_TYPE)"
                )
            }
        }

        override fun onDBCreate(db: SQLiteDatabase) {
            metaColumns.onDBCreate(db)
        }

        override fun onDBUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {

            if (newVersion < oldVersion) {
                try {
                    db.execSQL("drop table if exists $table")
                } catch (ex: Throwable) {
                    log.e(ex, "delete DB failed.")
                }
                metaColumns.onDBCreate(db)
                return
            }

            if (oldVersion < 2 && newVersion >= 2) {
                metaColumns.onDBCreate(db)
                return
            }

            if (oldVersion < 40 && newVersion >= 40) {
                try {
                    db.execSQL("drop index if exists ${table}_a")
                } catch (ex: Throwable) {
                    log.e(ex, "drop index failed.")
                }
            }
            metaColumns.onDBUpgrade(db, oldVersion, newVersion)
            if (oldVersion < 40 && newVersion >= 40) {
                try {
                    db.execSQL(
                        "create unique index if not exists ${table}_b on $table ($COL_ACCOUNT_DB_ID,$COL_NOTIFICATION_TYPE)"
                    )
                } catch (ex: Throwable) {
                    log.e(ex, "can't add index.")
                }
            }
        }

        private val cache =
            ConcurrentHashMap<Long, ConcurrentHashMap<String, NotificationTracking>>()

        private fun <K : Any, V : Any> ConcurrentHashMap<K, V>.getOrCreate(
            key: K,
            creator: () -> V,
        ): V {
            var v = this[key]
            if (v == null) v = creator().also { this[key] = it }
            return v
        }

        private fun loadCache(accountDbId: Long, notificationType: String): NotificationTracking? =
            cache[accountDbId]?.get(notificationType)

        private fun clearCache(accountDbId: Long, notificationType: String): NotificationTracking? =
            cache[accountDbId]?.remove(notificationType)

        private fun saveCache(
            accountDbId: Long,
            notificationType: String,
            nt: NotificationTracking,
        ) {
            cache.getOrCreate(accountDbId) {
                ConcurrentHashMap<String, NotificationTracking>()
            }[notificationType] = nt
        }

        /////////////////////////////////////////////////////////////////////////////////

        private const val WHERE_AID = "$COL_ACCOUNT_DB_ID=? and $COL_NOTIFICATION_TYPE=?"
    }

    fun toContentValues() = ContentValues().apply {
        put(COL_ACCOUNT_DB_ID, accountDbId)
        put(COL_NOTIFICATION_TYPE, notificationType)
        nid_read.putMayNull(this, COL_NID_READ)
        nid_show.putMayNull(this, COL_NID_SHOW)
        post_id.putMayNull(this, COL_POST_ID)
        put(COL_POST_TIME, post_time)
    }

    class Access(val db: SQLiteDatabase) {

        fun load(
            // ログ出力だけに使う
            acct: Acct,
            accountDbId: Long,
            notificationType: String,
        ): NotificationTracking {
            loadCache(accountDbId, notificationType)
                ?.takeIf { !it.dirty }
                ?.let {
                    log.d(
                        "${acct.pretty}/$notificationType load-cached. post=(${it.post_id},${it.post_time}), read=${it.nid_read}, show=${it.nid_show}"
                    )
                    return it
                }

            val whereArgs = arrayOf(accountDbId.toString(), notificationType)
            val dst = NotificationTracking()
            dst.accountDbId = accountDbId
            dst.notificationType = notificationType
            try {
                db.rawQuery(
                    "select * from $table where $COL_ACCOUNT_DB_ID=? and $COL_NOTIFICATION_TYPE=?",
                    whereArgs,
                )?.use { cursor ->
                    if (cursor.moveToFirst()) {
                        dst.id = cursor.getLong(COL_ID)

                        dst.post_id = EntityId.entityId(cursor, COL_POST_ID)
                        dst.post_time = cursor.getLong(COL_POST_TIME)

                        val show = EntityId.entityId(cursor, COL_NID_SHOW)
                        if (show == null) {
                            dst.nid_show = null
                            dst.nid_read = null
                        } else {
                            dst.nid_show = show
                            val read = EntityId.entityId(cursor, COL_NID_READ)
                            if (read == null) {
                                dst.nid_read = null
                            } else {
                                val r2 = minComparable(show, read)
                                dst.nid_read = r2
                                if (r2 != read) {
                                    log.i("$acct/$notificationType read>show! clip to $show")
                                    val cv = ContentValues()
                                    show.putTo(cv, COL_NID_READ) //変数名とキー名が異なるのに注意
                                    db.update(table, cv, WHERE_AID, whereArgs)
                                }
                            }
                        }

                        log.d(
                            "${acct.pretty}/$notificationType load. post=(${dst.post_id},${dst.post_time}), read=${dst.nid_read}, show=${dst.nid_show}"
                        )
                        saveCache(accountDbId, notificationType, dst)
                    }
                }
            } catch (ex: Throwable) {
                log.e(ex, "${acct.pretty} load failed.")
            } finally {
                dst.dirty = false
            }
            return dst
        }

        fun updateRead(accountDbId: Long, notificationType: String) {
            try {
                val whereArgs = arrayOf(accountDbId.toString(), notificationType)
                db.rawQuery(
                    "select $COL_NID_SHOW, $COL_NID_READ from $table where $COL_ACCOUNT_DB_ID=? and $COL_NOTIFICATION_TYPE=?",
                    whereArgs,
                )?.use { cursor ->
                    when {
                        !cursor.moveToFirst() -> log.e("updateRead[$accountDbId,$notificationType]: can't find the data row.")

                        else -> {
                            val nid_show = EntityId.entityId(cursor, COL_NID_SHOW)
                            val nid_read = EntityId.entityId(cursor, COL_NID_READ)
                            when {
                                nid_show == null ->
                                    log.e("updateRead[$accountDbId,$notificationType]: nid_show is null.")

                                nid_read != null && nid_read >= nid_show ->
                                    log.e("updateRead[$accountDbId,$notificationType]: nid_read already updated.")

                                else -> {
                                    log.i("updateRead[$accountDbId,$notificationType]: update nid_read as $nid_show...")
                                    val cv = ContentValues()
                                    nid_show.putTo(cv, COL_NID_READ) //変数名とキー名が異なるのに注意
                                    db.update(table, cv, WHERE_AID, whereArgs)
                                    clearCache(accountDbId, notificationType)
                                }
                            }
                        }
                    }
                }
            } catch (ex: Throwable) {
                log.e(ex, "updateRead[$accountDbId] failed.")
            }
        }

        fun resetPostAll() {
            try {
                val cv = ContentValues()
                cv.putNull(COL_POST_ID)
                cv.put(COL_POST_TIME, 0)
                db.update(table, cv, null, null)
                cache.clear()
            } catch (ex: Throwable) {
                log.e(ex, "resetPostAll failed.")
            }
        }

        // アカウント設定から手動で呼ばれる
        fun resetTrackingState(accountDbId: Long?) {
            accountDbId ?: return
            try {
                db.delete(table, "$COL_ACCOUNT_DB_ID=?", arrayOf(accountDbId.toString()))
                cache.remove(accountDbId)
            } catch (ex: Throwable) {
                log.e(ex, "resetTrackingState failed.")
            }
        }

        fun save(
            // ログ出力だけに使う
            acct: Acct,
            item: NotificationTracking,
        ) {
            try {
                if (!item.dirty) return
                val rowId = item.toContentValues()
                    .replaceTo(db, table)
                if (item.id == -1L) item.id = rowId

                log.d("${acct.pretty}/${item.notificationType} save. post=(${item.post_id},${item.post_time})")
                item.dirty = false
                clearCache(item.accountDbId, item.notificationType)
            } catch (ex: Throwable) {
                log.e(ex, "save failed.")
            }
        }

        fun updatePost(postId: EntityId, postTime: Long, item: NotificationTracking) {
            try {
                item.post_id = postId
                item.post_time = postTime
                if (!item.dirty) return
                val cv = item.toContentValues().apply {
                    postId.putTo(this, COL_POST_ID)
                    put(COL_POST_TIME, postTime)
                }
                val rows = db.update(
                    table,
                    cv,
                    WHERE_AID,
                    arrayOf(item.accountDbId.toString(), item.notificationType)
                )
                log.d("updatePost account_db_id=${item.accountDbId}, nt=${item.notificationType}, post=${postId},${postTime} update_rows=${rows}")
                item.dirty = false
                clearCache(item.accountDbId, item.notificationType)
            } catch (ex: Throwable) {
                log.e(ex, "updatePost failed.")
            }
        }
    }
}
